docs/advanced/signal_generators.md

```diff
@@ -0,0 +1,558 @@
+As such, it is up to the user to implement these checks.
+```
+
+Each class has children which inherit the base abstracted methods. These methods are tailored to each signal generator so the
+methods handle similarly, regardless of the different PI commands required.
+
+- `source_device_constants` is a property which holds information about what functions
+  and memory sizes are allowed.
+
+```{tip}
+`source_device_constants.functions` will provide an enum of possible functions to generate on the signal generator used.
+```
+
+`generate_function` is a method which allows for the user to request a function from
+any source channel provided an amplitude, frequency and offset is supplied. Other key features
+include the ability to manipulate specific aspects of certain functions. Ramp waveforms can have their symmetry changed
+and duty cycle can be altered for pulse functions. The termination of the IAFG and any AFG can be
+specified using `HIGHZ` or `FIFTY` string literals. If the output needs to be inverted,
+the polarity can be changed on AFGs.
+
+```{warning}
+`generate_function` allows function parameters which can exceed actual generation bounds.
+`get_waveform_constraints` should be used in tandem with `generate_function`, or utilizing the constraints provided in
+[Signal Generators](#signal-generators).
+```
+
+The `setup_burst` method places the source into a state for waveforms to be generated a set number
+of times. All parameters passed into the method are functionally identical to generate_function, besides burst_count,
+which specifies how many cycles of the waveform are to be generated.
+
+```{warning}
+`setup_burst` will set parameters which can effect the signal generators behavior. Changing these parameters
+will likely cause burst to stop functioning.
+```
+
+`generate_burst`  writes a trigger to the source, initiating the generation of a burst of waveforms.
+
+`get_waveform_constraints` will return a series of constraints that the signal generators must be within to be generated.
+These constraints can be used before generating a function to make sure that the parameters you will be supplying
+are not outside the bounds. The method only requires the function (except on AWGs) to be provided.
+However, different aspects may need to be provided to get a more accurate representation of what can actually be generated.
+If no other inputs are provided, the smallest range of the attribute is chosen.
+AWGs signal path effects the range of the offset and amplitude.
+Higher frequencies on AFGs will lower the upper bound of the amplitude,
+alongside the which impedance is set.
+
+`set_waveform_properties` is functionally identical to generate_function, but does not turn the channel
+off or on, nor will it stop or start an AWG.
+
+## Signal Generators
+
+An overview of the different signal generators which are covered when using `tm_devices`.
```

Comment:
```suggestion
An overview of the different signal generators which are available in `tm_devices`.
```

Reply:
This summary should also indicate that the constraints are provided for each model.